lib/workers/repository/update/branch/index.ts

```diff
@@ -324,13 +324,14 @@ export async function processBranch(
         const currentVersion = upgrade.currentVersion!;
         const newVersion = upgrade.newVersion!;
         if (isActiveConfidenceLevel(minimumConfidence)) {
-          const confidence = await getMergeConfidenceLevel(
-            datasource,
-            depName,
-            currentVersion,
-            newVersion,
-            updateType
-          );
+          const confidence =
+            (await getMergeConfidenceLevel(
+              datasource,
+              depName,
+              currentVersion,
+              newVersion,
+              updateType
+            )) ?? 'neutral';
```

Comment:
the http request is handled in a try catch block, therefore if it rejects in is handled in place and `undefined` will return, previously `neutral`.

overall, `undefined` will be returned if:
- no/invalid token
- `isMergeConfidence(res.body.confidence)` = false
- http request rejected

Reply:
If for say, we group deps by merge confidence and we already have a PR for it, it will be updated on every run if needed. 

i.e. if we have API failure and we couldn't get no merge confidence at all, the grouped PR will be auto closed. if some deps have its mc changed, the PR will be updated accordingly etc.

Reply:
If for say, we group deps by merge confidence and we already have a PR for it, it will be updated on every run if needed. 

i.e. if we have API failure and we couldn't get no merge confidence at all, the grouped PR will be auto closed. if some deps have its mc changed, the PR will be updated accordingly etc.

Reply:
We need to consider external host errors to abort.

If they configure matchConfidence but have no credentials then I think we should use neutral and continue, plus warn them if their config mistake. 

If they have a wrong token or unreachable API then we should consider aborting the run

Reply:
so abort run on 5xx? 

if we return `neutral` it might be out of sync with the PR badges within the grouped PR, is that ok?

Reply:
Note that if we return `neutral` when no token is set, every package will have its mc set to `neutral`, even for users that don't have the mc matcher set.

As a result, the `PackageFiles with updates` log message will be filled with `neutral` which might be misleading and possibly cause users to miss-interpret the benefits of mc.

I suppose we can check within `getMergeConfidenceLevel` if either of `matchConfidence` or `minConfidence` are enabled but that would be plain ugly (and might even be blown out of proportion if we plan to use this function in other use cases).

Reply:
We don't need to query for mergeConfidence unless we have a matchConfidence set? I guess it's ok to leave confidence as undefined as long as we later treat that as neutral. And yes if we have a timeout or other failure of the API I think we need to abort so that it doesn't cause major flapping of PRs

Reply:
> We don't need to query for mergeConfidence unless we have a matchConfidence set?

The `getConfidenceLevel` function is being called regardless of the package rules matcher. 
it is being called as part of the lookup phase and while building the `packageFiles` object, that's why I think it would be best to return `undefined` if there is no token set (which both `minMergeConfidence` and `matchConfidence` requires). 


So currently (this PR) we will return `undefined` only when there is no API key set. 
in this case -
- `minMergeConfidence` will fallback outside the function to `neutral` 
- deps will not have any merge confidence values -> `package rules` wont be able to match - as expected

if we do have a token set -
- on 403/any server side error we abort
- on any other client side error, we fallback to `neutral` 
- otherwise, we got the appropriate mc value from the api

Reply:
I had been thinking that the token could be configured low level with hostRules, but in that case it would be a little inelegant to 'cancel' the request at a low level and return. Did you have any alternative ideas, perhaps similar to how we actively check for a github.com token?

Reply:
Now we are checking the token at the global init phase, just after host rules are initialized and at the same phase the `github` token is checked (`initPlatform`).

wdyt?

Reply:
@rarkins 
Is this https://github.com/Gabriel-Ladzaretti/renovate/pull/11/commits/1d76113528e76a614ea257de14eae18b0a3f62f5 any closer to what you had in mind?